### PR TITLE
Add Dockerfile and ghidra_deploy.sh

### PR DESCRIPTION
**Pull request by skyblue302**

### CODE REVIEW

File: deployment-scripts/ghidra_deploy/ghidra_deploy.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+# Set non-interactive installation mode
+export DEBIAN_FRONTEND=noninteractive
+
+echo "ghidrauser:ghidrauser" | sudo chpasswd
+export SUDO_PASS="ghidrauser"
```
Comment (louiskyee) [inline deployment-scripts/ghidra_deploy/ghidra_deploy.sh:7]: Problem: The password is hardcoded in the script, which poses a security risk.
Suggestion: Use a variable to store the password and ensure the script is stored in a secure location. Consider reading the password from a secure location, such as an environment variable.

File: deployment-scripts/ghidra_deploy/ghidra_deploy.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+# Set non-interactive installation mode
+export DEBIAN_FRONTEND=noninteractive
+
+echo "ghidrauser:ghidrauser" | sudo chpasswd
```
Comment (bolin8017) [inline deployment-scripts/ghidra_deploy/ghidra_deploy.sh]: Problem: The password is hardcoded in the script, which poses a security risk.
Suggestion: Use a variable to store the password and ensure the script is stored in a secure location. Consider reading the password from a secure location, such as an environment variable.

File: deployment-scripts/ghidra_deploy/ghidra_deploy.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+export SUDO_PASS="ghidrauser"
+echo $SUDO_PASS | sudo -S apt-get update
+echo $SUDO_PASS | sudo -S apt-get install -y git make wget unzip python3 python3-pip gcc g++ python3-venv python-is-python3 tar
```
Comment (bolin8017) [inline deployment-scripts/ghidra_deploy/ghidra_deploy.sh]: Problem: The command echo $SUDO_PASS | sudo -S is repeated multiple times, which is not efficient.
Suggestion: Define a function to run commands with sudo and use it throughout the script.

File: deployment-scripts/ghidra_deploy/ghidra_deploy.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+if [[ "$JAVA_HOME" =~ .*jdk-17.* ]]; then
+    echo "JAVA_HOME is correctly set to JDK 17."
+else
+    echo "JAVA_HOME was not set correctly. Attempting to set JAVA_HOME..."
+    JAVA_HOME=$(update-alternatives --list javac | grep java-17 | sed 's:/bin/javac::' | head -n 1)
+    export JAVA_HOME
+fi
+
+echo "JAVA_HOME set to $JAVA_HOME"
```
Comment (bolin8017) [inline deployment-scripts/ghidra_deploy/ghidra_deploy.sh]: Problem: The JAVA_HOME setup is repeated and uses two different methods.
Suggestion: Combine the setup steps and use a consistent method.

File: deployment-scripts/ghidra_deploy/ghidra_deploy.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+tar -xzf /home/ghidrauser/tmp/jdk-17.0.7_linux-x64_bin.tar.gz -C /home/ghidrauser/jdk-17 --strip-components=1
+
+export JAVA_HOME=/home/ghidrauser/jdk-17
+export PATH=$JAVA_HOME/bin:$PATH
+
+# Check if java is set correctly
+# java -version
+
+# Set JAVA_HOME to the path of JDK 17
+export JAVA_HOME=$(dirname $(dirname $(readlink -f $(which javac))))
```
Comment (bolin8017) [inline deployment-scripts/ghidra_deploy/ghidra_deploy.sh]: Problem: Environment variables should be persisted in the .profile file to ensure they are available in new shell sessions.
Suggestion: Consolidate the commands to append to the .profile file.

**Review by bolin8017 — CHANGES_REQUESTED** [review]
Here is the major issues and suggestions for the `ghidra_deploy.sh` script:

1. **Hardcoded Password**
   - **Problem**: The password is hardcoded in the script, posing a security risk.
   - **Suggestion**: Use a variable to store the password and ensure the script is stored in a secure location. Consider reading the password from a secure location, such as an environment variable.

   ```bash
   USER="ghidrauser"
   PASS="ghidrauser"
   echo "$USER:$PASS" | sudo chpasswd
   ```

2. **Repeated sudo Password Input**
   - **Problem**: The command `echo $SUDO_PASS | sudo -S` is repeated multiple times, which is inefficient.
   - **Suggestion**: Define a function to run commands with sudo and use it throughout the script.

   ```bash
   run_sudo() {
       echo $PASS | sudo -S "$@"
   }
   run_sudo apt-get update
   run_sudo apt-get install -y git make wget unzip python3 python3-pip gcc g++ python3-venv python-is-python3 tar
   ```

3. **Repeated JAVA_HOME Setup**
   - **Problem**: The `JAVA_HOME` setup is repeated and uses two different methods.
   - **Suggestion**: Combine the setup steps and use a consistent method.

   ```bash
   export JAVA_HOME=/home/$USER/jdk-17
   export PATH=$JAVA_HOME/bin:$PATH
   ```

4. **Lack of Error Handling**
   - **Problem**: The script lacks error handling for download and installation steps, which might continue execution even if a step fails.
   - **Suggestion**: Add error checking after each critical step to ensure successful execution.

   ```bash
   wget https://download.oracle.com/java/17/archive/jdk-17.0.7_linux-x64_bin.tar.gz -P /home/$USER/tmp || { echo "Failed to download JDK"; exit 1; }
   ```

5. **Persisting Environment Variables**
   - **Problem**: Environment variables should be persisted in the `.profile` file to ensure they are available in new shell sessions.
   - **Suggestion**: Consolidate the commands to append to the `.profile` file.

   ```bash
   echo -e "export JAVA_HOME=$JAVA_HOME\nexport GRADLE_HOME=$GRADLE_HOME\nexport PATH=\$JAVA_HOME/bin:\$GRADLE_HOME/bin:\$PATH" >> /home/$USER/.profile
   source /home/$USER/.profile
   ```

6. **GHIDRA_HOME Environment Variable Setup**
   - **Problem**: The setup for `GHIDRA_HOME` should be consistent with previous setups and check if it is set correctly.
   - **Suggestion**: Combine environment variable setup steps.

   ```bash
   export GHIDRA_HOME=/home/$USER/ghidra/ghidra_10.3.1_PUBLIC
   ```

7. **Repeated Download and Extraction**
   - **Problem**: The download and extraction steps lack error handling and should use variables to simplify paths.
   - **Suggestion**: Add error handling and use variables for paths.

   ```bash
   GHIDRA_ARCHIVE=/home/$USER/tmp/ghidra_10.3.1_PUBLIC_20230614.zip
   wget https://github.com/NationalSecurityAgency/ghidra/releases/download/Ghidra_10.3.1_build/ghidra_10.3.1_PUBLIC_20230614.zip -P /home/$USER/tmp || { echo "Failed to download Ghidra"; exit 1; }
   unzip $GHIDRA_ARCHIVE -d /home/$USER/ghidra || { echo "Failed to unzip Ghidra"; exit 1; }
   ```

8. **Activating Virtual Environment**
   - **Problem**: Creating and activating the virtual environment lacks error handling.
   - **Suggestion**: Check if the virtual environment is created and activate it accordingly.

   ```bash
   python3 -m venv $GHIDRA_HOME/venv || { echo "Failed to create virtual environment"; exit 1; }
   if [ -f "$GHIDRA_HOME/venv/bin/activate" ]; then
       source $GHIDRA_HOME/venv/bin/activate
   else
       echo "Virtual environment activation script not found."
       exit 1
   fi
   ```

9. **Missing Cleanup Step**
   - **Problem**: The script does not clean up temporary files after execution.
   - **Suggestion**: Add a cleanup step at the end of the script.

   ```bash
   # Clean up temporary files
   rm -rf /home/$USER/tmp
   ```

**Review by bolin8017 — APPROVED** [review]
Thank you for updating this Ghidra deployment file and resolving the issues from the previous PR. I will merge this file into the main branch. I greatly appreciate your kind assistance.

**Pull request merged**